### PR TITLE
Bug/fix broken backup feature 😠 🤪 

### PR DESCRIPTION
**Pull request by norkator**
Should look line 140 at BackupUtils.java..
getDBFileInputSteamFromExternalStorage

Spent hours now trying to figure this new file access permissions system out at api 29 but could not find solution yet so did "easy" hack which doesn't solve this for long => android:requestLegacyExternalStorage="true"

Maybe someone else knows how to get access properly to external storage... like download folder to read file.

@developerfromjokela if you see this and know something, throw comment! 👍

### CODE REVIEW

**Comment by ristomatti** [comment]
This one looks like something I'd actually need to test. At least today I won't be setting up any new dev environments anymore... Tomorrow I'm heading to summer cottage for a week and I'll be relying on my laptop screen making code reviews somewhat more cumbersome. But let's see!

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -32,7 +32,8 @@
         android:theme="@style/AppTheme"
         android:usesCleartextTraffic="true"
         tools:ignore="GoogleAppIndexingWarning"
-        tools:targetApi="m">
+        tools:targetApi="m"
```
Comment (ristomatti) [inline app/src/main/AndroidManifest.xml:35]: Will this break support for older Android versions than 6.0?

File: app/src/main/java/com/nitramite/paketinseuranta/DatabaseHelper.java [hunk 2]
```diff
@@ -64,7 +57,7 @@
 
 
     // DATABASE NAME
-    private static final String DATABASE_NAME = "PARCELS.db";
+    public static final String DATABASE_NAME = "PARCELS.db";
```
Comment (ristomatti) [inline app/src/main/java/com/nitramite/paketinseuranta/DatabaseHelper.java:60]: Move this into `Constants.java` instead?

File: app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (norkator) [reply app/src/main/AndroidManifest.xml:35]: "Indicates that Lint should treat this type as targeting a given API level, no matter what the project target is"
=> No affect anywhere. I should by the way delete this completely, it's some old stuff.

Comment (norkator) [reply app/src/main/AndroidManifest.xml]: https://github.com/norkator/paketin-seuranta/pull/35/commits/bc8b05010d741deee5de89325eb30ece76c92e23
Deleted it.

Comment (norkator) [reply app/src/main/AndroidManifest.xml]: @ristomatti one other note.. this app has not supported < Android 6.0 (API level 23) for long time.

minSdkVersion 23
targetSdkVersion 29

File: app/src/main/java/com/nitramite/utils/BackupUtils.java [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    public static Backup backupDatabase(Context context) {
+        Backup backup = new Backup();
+        try {
+            FileInputStream dbFileInputStream = getDBFileInputStream(context, backup);
+            OutputStream dbFileOutputStream = getDBOutputStream(context, backup);
+
+            // Write new file out
+            byte[] buffer = new byte[1024];
+            int length;
+            while ((length = dbFileInputStream.read(buffer)) > 0) {
```
Comment (ristomatti) [inline app/src/main/java/com/nitramite/utils/BackupUtils.java:42]: Interesting syntax, I don't think I've never seen an assignment inside a while loop on Java. :open_mouth:

File: app/src/main/java/com/nitramite/utils/BackupUtils.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+import java.util.Objects;
+
+import static com.nitramite.paketinseuranta.DatabaseHelper.DATABASE_NAME;
+
+public class BackupUtils {
+
+    @NonNls
+    private static final String TAG = "BackupUtils";
+
+    // Backup database
```
Comment (ristomatti) [inline app/src/main/java/com/nitramite/utils/BackupUtils.java:32]: Remove this useless comment.

File: app/src/main/java/com/nitramite/utils/BackupUtils.java [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+            backup.setSuccess(true);
+            return backup;
+        } catch (IOException e) {
+            return backup;
+        }
+    }
+
+
+    // Restore database
```
Comment (ristomatti) [inline app/src/main/java/com/nitramite/utils/BackupUtils.java:57]: Remove this useless/redundant comment.

File: app/src/main/java/com/nitramite/utils/BackupUtils.java [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+            Log.i(TAG, targetPathFile.getPath());
+
+            OutputStream output = new FileOutputStream(targetPathFile);
+            byte[] buffer = new byte[1024];
+            int length;
+            while ((length = fileInputStream.read(buffer)) > 0) {
+                output.write(buffer, 0, length);
+            }
+
+            // Flush & Close
```
Comment (ristomatti) [inline app/src/main/java/com/nitramite/utils/BackupUtils.java:73]: Redundant comment...

File: app/src/main/java/com/nitramite/utils/BackupUtils.java [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+     * @return file input steam
+     * @throws FileNotFoundException db file not found
+     */
+    private static FileInputStream getDBFileInputSteamFromExternalStorage(Context context) throws FileNotFoundException, NullPointerException {
+        FileInputStream fileInputStream = null;
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+
+            // TODO: Wrong implementation because Google's documentation sucks with this subject.
+            // TODO: currently using android:requestLegacyExternalStorage="true" at manifest.
+            // TODO: research proper solution!
```
Comment (ristomatti) [inline app/src/main/java/com/nitramite/utils/BackupUtils.java:143]: Adding a separate `TODO` for each line will generate more todo's in the IDE's list: :stuck_out_tongue:

File: app/src/main/java/com/nitramite/utils/BackupUtils.java [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+            @SuppressWarnings("deprecation") String downloadsDir = Environment.getExternalStorageDirectory() + "/Download/";
```
Comment (ristomatti) [inline app/src/main/java/com/nitramite/utils/BackupUtils.java:144]: Will the dir be configurable / is this a default value or a fixed one? Make it a constant `BACKUP_PATH` inside `Contants.java`?

File: app/src/main/res/values-fi-rFI/strings.xml [hunk 9]
```diff
@@ -258,7 +256,11 @@
     <string name="description">Selite</string>
     <string name="carrier_detector_tool">Kuljetusyhtiön tunnistustyökalu</string>
     <string name="carrier_detector_tool_description">Tämä työkalu yrittää tunnistaa pakettia eri toimitusyhtiöistä koodia hakemalla. Voit lopuksi valita sopivan yhtiön jos alle ilmestyy valintoja.</string>
-    <string name="reference_to_database_is_null_cannot_add_items">Referenssi tietokantaan puuttuu, pakettien lisäys ei onnistu!</string>
+	<string name="reference_to_database_is_null_cannot_add_items">Referenssi tietokantaan puuttuu, pakettien lisäys ei onnistu!</string>
+    <string name="main_menu_taking_backup_success">Varmuuskopion otto onnistui. Tiedosto löytyy sijainnista:</string>
```
Comment (ristomatti) [inline app/src/main/res/values-fi-rFI/strings.xml:260]: "Varmuuskopion otto" -> "Varmuuskopiointi"
"Tiedosto löytyy polusta:"

File: app/src/main/res/values-fi-rFI/strings.xml [hunk 10]
```diff
@@ -258,7 +256,11 @@
...
+    <string name="main_menu_taking_backup_failed">Varmuuskopion otto epäonnistui. Yritä kerran uudelleen. Jos varmuuskopion otto epäonnistuu uudelleen, ota yhteys kehittäjään.</string>
```
Comment (ristomatti) [inline app/src/main/res/values-fi-rFI/strings.xml:261]: "Varmuuskopioint epäonnistui, yritä uudelleen. Jos varmuuskopiointi edelleen epäonnistuu, ota yhteys kehittäjään."

File: app/src/main/res/values-fi-rFI/strings.xml [hunk 11]
```diff
@@ -258,7 +256,11 @@
...
+    <string name="backup_location_external_storage_downloads_dir">ulkoisen muistin lataukset kansio, tiedosto nimeltä </string>
```
Comment (ristomatti) [inline app/src/main/res/values-fi-rFI/strings.xml:262]: "ulkoisen tallennustilan latauskansio, tiedosto xxx"

File: app/src/main/res/values/strings.xml [hunk 12]
```diff
@@ -259,6 +257,10 @@
     <string name="carrier_detector_tool">Carrier detector tool</string>
     <string name="carrier_detector_tool_description">This tool tries to find possible carriers for given tracking code. If tool gives options, you can select one via clicking carrier image.</string>
     <string name="reference_to_database_is_null_cannot_add_items">Reference to database is missing, cannot add items!</string>
+    <string name="main_menu_taking_backup_success">Taking backup was successful. File can be found from: </string>
```
Comment (ristomatti) [inline app/src/main/res/values/strings.xml:260]: "Backup was succesful. Backup file can be found from: "

File: app/src/main/res/values/strings.xml [hunk 13]
```diff
@@ -259,6 +257,10 @@
...
+    <string name="main_menu_taking_backup_failed">Taking backup failed. Please try again. If taking backup fails multiple time, please contact developer.</string>
```
Comment (ristomatti) [inline app/src/main/res/values/strings.xml:261]: "Backup failed, please try again. Please contact the developer if backup still fails."

File: app/src/main/res/values/strings.xml [hunk 14]
```diff
@@ -259,6 +257,10 @@
...
+    <string name="backup_location_external_storage_downloads_dir">external storage downloads directory with a name </string>
```
Comment (ristomatti) [inline app/src/main/res/values/strings.xml:262]: "external storage downloads directory named "

**Comment by developerfromjokela** [comment]
What if we use the application's "data" folder (public one)? It's located at `<internal_storage>/android/data/<package_name>`. Only small downsite is that basic user wouldn't likely find this folder. Maybe putting a button to open the path will help?

And there's one thing about that workaround in Manifest, if you're in target SDK 30 or newer, it will ignore that flag. And setting to SDK 29 isn't a longtime solution, Google will force us to use newer one.

I saw at one app, it was an Screen recorder, when it launched, it asked user to select the save path. And it opened Android built-in screen for selecting folder. That will grant access to the selected folder. 

As far as I know that is the right way to fix this issue.

File: app/src/main/java/com/nitramite/paketinseuranta/DatabaseHelper.java (hunk 2 above)
Comment (norkator) [reply app/src/main/java/com/nitramite/paketinseuranta/DatabaseHelper.java:60]: Yes not bad idea

**Comment by norkator** [comment]
> What if we use the application's "data" folder (public one)? It's located at `<internal_storage>/android/data/<package_name>`. Only small downsite is that basic user wouldn't likely find this folder. Maybe putting a button to open the path will help?
> 
> And there's one thing about that workaround in Manifest, if you're in target SDK 30 or newer, it will ignore that flag. And setting to SDK 29 isn't a longtime solution, Google will force us to use newer one.
> 
> I saw at one app, it was an Screen recorder, when it launched, it asked user to select the save path. And it opened Android built-in screen for selecting folder. That will grant access to the selected folder.
> 
> As far as I know that is the right way to fix this issue.

Thanks for these details, need to dig into this.

File: app/src/main/java/com/nitramite/paketinseuranta/DatabaseHelper.java (hunk 2 above)
Comment (norkator) [reply app/src/main/java/com/nitramite/paketinseuranta/DatabaseHelper.java]: https://github.com/norkator/paketin-seuranta/pull/35/commits/027415733094de84690c8c022c70e2213e0fce62

File: app/src/main/java/com/nitramite/utils/BackupUtils.java (hunk 3 above)
Comment (norkator) [reply app/src/main/java/com/nitramite/utils/BackupUtils.java:42]: Old code.. originating to Stackoverflow where you see everything you cannot imagine.

File: app/src/main/java/com/nitramite/utils/BackupUtils.java (hunk 4 above)
Comment (norkator) [reply app/src/main/java/com/nitramite/utils/BackupUtils.java:32]: Modified to be proper comments

File: app/src/main/java/com/nitramite/utils/BackupUtils.java (hunk 5 above)
Comment (norkator) [reply app/src/main/java/com/nitramite/utils/BackupUtils.java:57]: Modified to be proper comments

File: app/src/main/java/com/nitramite/utils/BackupUtils.java (hunk 7 above)
Comment (norkator) [reply app/src/main/java/com/nitramite/utils/BackupUtils.java:143]: Have not been doing multiline todo before... fixed to be one TODO still three lines.

File: app/src/main/java/com/nitramite/utils/BackupUtils.java (hunk 6 above)
Comment (norkator) [reply app/src/main/java/com/nitramite/utils/BackupUtils.java:73]: Removed

File: app/src/main/res/values-fi-rFI/strings.xml (hunk 9 above)
Comment (norkator) [reply app/src/main/res/values-fi-rFI/strings.xml:260]: Korjattu

File: app/src/main/java/com/nitramite/utils/BackupUtils.java (hunk 8 above)
Comment (norkator) [reply app/src/main/java/com/nitramite/utils/BackupUtils.java:144]: I let this be open, @developerfromjokela comment would change this to be dynamic.

File: app/src/main/res/values-fi-rFI/strings.xml (hunk 10 above)
Comment (norkator) [reply app/src/main/res/values-fi-rFI/strings.xml:261]: i puuttui mutta laitoin tuon muuten. Parempi.

File: app/src/main/res/values-fi-rFI/strings.xml (hunk 11 above)
Comment (norkator) [reply app/src/main/res/values-fi-rFI/strings.xml:262]: Korjattu

File: app/src/main/res/values/strings.xml (hunk 12 above)
Comment (norkator) [reply app/src/main/res/values/strings.xml:260]: Ok

File: app/src/main/res/values/strings.xml (hunk 13 above)
Comment (norkator) [reply app/src/main/res/values/strings.xml:261]: Ok

File: app/src/main/res/values/strings.xml (hunk 14 above)
Comment (norkator) [reply app/src/main/res/values/strings.xml:262]: ok

**Comment by developerfromjokela** [comment]
@norkator Have you looked to this article: [https://medium.com/better-programming/all-you-need-to-know-about-scoped-storage-in-android-10-e621f40bc8b9](https://medium.com/better-programming/all-you-need-to-know-about-scoped-storage-in-android-10-e621f40bc8b9)

**Comment by developerfromjokela** [comment]
Here's an example: https://github.com/pg598595/ScopedStorageDemo

**Comment by norkator** [comment]
Thanks, I'll check/continue this today later

**Comment by norkator** [comment]
image

Slowly progressing

**Comment by norkator** [comment]
image

Still cannot get output stream with this part. I don't seem to understand this. I give up for now since this is really starting to piss me off.

**Comment by norkator** [comment]
There's some structures ready and base view **if someone wants to continue** I don't give a fuck.

**Comment by norkator** [comment]
Okay I try this again when time firing up that example app and testing with it first.

**Comment by norkator** [comment]
Removed that android:requestLegacyExternalStorage="true"
Taking backup works but its been taken to following location on api 29+
image

So as @developerfromjokela said I should make some sort of button which opens this location to folder finder and also make backup view nicer. Any recommendations how to improve visuals? @developerfromjokela @ristomatti

**Comment by developerfromjokela** [comment]
@norkator Would it be good to place a button named "Open backup folder" near the Close button?

**Comment by norkator** [comment]
> @norkator Would it be good to place a button named "Open backup folder" near the Close button?

You mean on this backup manager activity like bottom corner?

**Comment by developerfromjokela** [comment]
No I mean that dialog that you showed on that image

**Comment by norkator** [comment]
> No I mean that dialog that you showed on that image

Haha yeah indeed

**Comment by ristomatti** [comment]
Yep button there sounds like a good plan to me also. Is there a general activity that will then prompt the user which file manager app to use for opening the folder?

**Comment by norkator** [comment]
image

**Good**
* View looks like this. Taking backup and restoring backup works both >API29< and tested with different virtual and physical devices.
* Timed backup is on own view section and it also shows when last backup is taken.

image

**Bad**
* There is open location button on success dialog but no matter what I try for intents and after many stackoverflow pages file browser intent wont never open that file location on any device SDK level. I tried with right mime type, only folder without file and extension on path. Everything.

@developerfromjokela can you find solution for this or @ristomatti  ?

With using "backup" object you can pass any needed details from "BackupUtils.java" to void named "openFolderLocation" which currently has some random intent.

**Pull request merged**